src/firstpass.rs

```diff
@@ -1183,6 +1217,120 @@ impl<'a, 'b> FirstPass<'a, 'b> {
+    ///
+    /// If `trailing_attribute_block` of the return value is `Some(_)`, the
+    /// string starts with `{` and ends with `}`.
+    ///
+    /// If `end_ix` is less than or equal to `begin_ix`, the given input is
+    /// considered as empty.
+    ///
+    /// The closing `}` of an attribute block can have trailing whitespaces.
+    ///
+    /// This method does not trim the trailing whitespaces. It is callers'
+    /// responsibility to trim them if necessary.
+    fn extract_attribute_block_from_header_text(
+        &self,
+        begin_ix: usize,
+        end_ix: usize,
+    ) -> (usize, Option<&'a str>) {
+        if begin_ix >= end_ix {
+            return (end_ix, None);
+        }
+
+        let bytes = self.text.as_bytes();
+        let mut ix = end_ix;
+        ix -= scan_rev_while(&bytes[begin_ix..ix], |b| {
+            b == b'\n' || b == b'\r' || b == b' ' || b == b'\t'
+        });
+
+        let attr_block_close = ix - 1;
+        if bytes.get(attr_block_close) != Some(&b'}') {
+            // The last character is not `}`. No attribute blocks found.
+            return (end_ix, None);
+        }
+        // move cursor before the closing right brace (`}`)
+        ix -= 1;
+
+        ix -= scan_rev_while(&bytes[begin_ix..ix], |b| {
+            // Characters to be excluded:
+            //  * `{` and `}`: special characters to open and close an attribute block.
+            //  * `\\`: a special character to escape many characters and disable some syntaxes.
+            //      + Handling of this escape character differs among markdown processors.
+            //      + Escaped characters will be separate text node from neighbors, so
+            //        it is not easy to handle unescaped string and trim the trailing block.
+            //  * `<` and `>`: special characters to start and end HTML tag.
+            //      + No known processors converts `{#<i>foo</i>}` into
+            //        `id="&lt;i&gt;foo&lt;/&gt;"` as of this writing, so hopefully
+            //        this restriction won't cause compatibility issues.
+            //  * `\n` and `\r`: a newline character.
+            //      + Setext heading can have multiple lines. However it is hard to support
+            //        attribute blocks that have newline inside, since the parsing proceeds line by
+            //        line and lines will be separate nodes even they are logically a single text.
+            !matches!(b, b'{' | b'}' | b'<' | b'>' | b'\\' | b'\n' | b'\r')
+        });
+        if ix == begin_ix {
+            // `{` is not found. No attribute blocks available.
+            return (end_ix, None);
+        }
+        let attr_block_open = ix - 1;
+        if bytes[attr_block_open] != b'{' {
+            // `{` is not found. No attribute blocks available.
+            return (end_ix, None);
+        }
+
+        (
+            attr_block_open,
+            Some(&self.text[attr_block_open..=attr_block_close]),
```

Comment:
Okay so this is a completely normal operation, but in pulldown we try to limit the number of string slices that we take. It is a fairly costly operations, and usually we operate on bytes anyway. Could you return a byte slice instead?

Reply:
Returning byte slice here is possible, but `parse_inside_attribute_block()` needs a substring to get ID fragment and class names as `&str`.
I'll return `std::ops::Range<usize>`, as it can be useable to slice both string and bytes, and manipulating Range itself would be costless.